### PR TITLE
Key memory management for compound hash tables.

### PR DESCRIPTION
**Pull request by akuzm**
Some aggregation methods initially emplace a temporary StringRef key
into a hash table. Then, if the key was not seen before, they make a
persistent copy of the key and update the hash table with it. This
approach is not suitable for compound hash tables, because the logic of
when the persistent key is needed is more complex, and is contained
within the hash table itself.

In this commit, we switch to managing key memory with callbacks passed
to the hash table, that allow it to request a persistent copy of the key
if it is needed. This should be more appropriate for compound hash
tables.

This PR prepares for StringHashMap PR #5417.

Category (leave one):
- Performance Improvement
(Not sure we should write about it in the changelog, it is a refactoring PR. But maybe it has some performance benefit, let's look at the tests.)

### CODE REVIEW

**Comment by akuzm** [comment]
@amosbird @alexey-milovidov @KochetovNicolai I'll be glad to hear your thoughts on this.

File: dbms/src/Common/ColumnsHashing.h [hunk 1]
```diff
@@ -317,9 +316,9 @@ struct HashMethodSingleLowCardinalityColumn : public SingleColumnMethod
     }
 
     /// Get the key from the key columns for insertion into the hash table.
```
Comment (amosbird) [inline dbms/src/Common/ColumnsHashing.h]: comment key -> key holder

File: dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.h [hunk 2]
```diff
@@ -241,15 +236,12 @@ class AggregateFunctionGroupUniqArrayGeneric
...
-                    it->getValueMutable().data = arena->insert(it->getValue().data, it->getValue().size);
-                else
-                    it->getValueMutable().data = nullptr;
-            }
+                return;
+
+            // We have to copy the keys to our arena.
+            assert(arena != nullptr);
+            cur_set.emplaceKeyHolder(ArenaKeyHolder{rhs_elem.getValue(), *arena},
+                               it, inserted);
```
Comment (amosbird) [inline dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.h]: weird line wrap and indent. The line is short enough.

File: dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.h [hunk 3]
```diff
@@ -154,9 +150,21 @@ class AggregateFunctionGroupUniqArrayGeneric
...
+    static KeyHolder getKeyHolder(const IColumn & column, size_t row_num, Arena & arena)
     {
-        return getSerializationImpl<is_plain_column>(column, row_num, arena);
+        if constexpr (is_plain_column)
+        {
+            return ArenaKeyHolder{column.getDataAt(row_num), arena};
+        }
+        else
+        {
+            const char * begin = nullptr;
```
Comment (amosbird) [inline dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.h]: we can leave out this var

**Review by amosbird — COMMENTED** [review]
The interface looks great. Now we don't need onNewKey/onExistingKey scattering around. Will the compiler be smart enough to get rid of all the potential copies when constructing KeyHolders?

**Comment by akuzm** [comment]
> Will the compiler be smart enough to get rid of all the potential copies when constructing KeyHolders?

Hope so, but didn't check yet. If we see extra copying, we can change the interface slightly: pass keys as usual, and only pass wrappers when we have temporary StringRef keys. Then, make persistKey()/discardKey() overloaded template functions instead of methods, with default implementation that does nothing. I acually tried a similar thing with passing plain key pointers instead of NoopKeyHolder, but there were problems around getKeyHolder in hashing method, so I stopped. Not sure, maybe I have to try it again. The interface is going to be simpler (no NoopKeyHolders everywhere) and we won't have to worry about copying. I'll rest from this for a couple of days and will think again.

**Comment by akuzm** [comment]
`SELECT number % 10000000 AS k, count() FROM system.numbers GROUP BY k` is failing with x0.88 slowdown.

OTOH, another query shows x1.26 speedup: `SELECT URL, count() AS PageViews FROM hits_100m_single WHERE CounterID = 34 AND EventDate >= '2013-07-01' AND EventDate <= '2013-07-31' AND NOT Refresh AND IsLink AND NOT IsDownload GROUP BY URL ORDER BY PageViews DESC LIMIT 1000`

Will have to investigate them in detail using the new comparison mode in performance tests.

File: dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.h (hunk 3 above)
Comment (akuzm) [reply dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.h]: It's an out-parameter, so we have to keep it.

File: dbms/src/Common/HashTable/HashTable.h [hunk 4]
```diff
@@ -630,6 +631,8 @@ class HashTable :
 
 
     /// If the key is zero, insert it into a special place and return true.
+    /// We don't have to persist a zero key, because it's not actually inserted.
+    /// That's why we just take a Key by value, an not a smart pointer to it.
```
Comment (akuzm) [inline dbms/src/Common/HashTable/HashTable.h]:
```suggestion
    /// That's why we just take a Key by value, an not a key holder.
```

File: dbms/src/Common/HashTable/HashTable.h [hunk 5]
```diff
@@ -739,19 +752,20 @@ class HashTable :
...
     {
-        size_t hash_value = hash(x);
-        if (!emplaceIfZero(x, it, inserted, hash_value))
-            emplaceNonZero(x, it, inserted, hash_value);
+        const auto & key = keyHolderGetKey(key_holder);
+        emplace(key_holder, it, inserted, hash(key));
     }
 
-
-    /// Same, but with a precalculated value of hash function.
```
Comment (akuzm) [inline dbms/src/Common/HashTable/HashTable.h]: Put the comment back, add a comment about key holders.

File: dbms/src/Common/HashTable/HashTableKeyHolder.h [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  * different key holders and corresponding functions for storing StringRef
+  * keys to Arena.
+  */
+
+/**
+  * Returns the key. Can return the temporary key initially.
+  * After the call to keyHolderPersistKey(), must return the persistent key.
+  */
+template <typename Key>
+inline Key & ALWAYS_INLINE keyHolderGetKey(Key && key) { return key; }
```
Comment (alexey-milovidov) [inline dbms/src/Common/HashTable/HashTableKeyHolder.h]: Maybe it won't work if we pass temporary. `Key &` will collapse to reference to temporary.
Better to use `std::forward`?

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Performance test is clear.

**Pull request merged**